src/XMakeBuildEngine/Construction/Solution/SolutionProjectGenerator.cs

```diff
@@ -1942,6 +1917,13 @@ private void AddTraversalTargetForProject(ProjectInstance traversalProject, Proj
                 actualTargetName += ":" + targetToBuild;
             }
 
+            // Don't add the target again.  The user might have specified /t:Project:target which was already added but only this method knows about Project:Target so
+            // after coming up with that target name, it can check if it has already been added.
+            if (traversalProject.Targets.ContainsKey(actualTargetName))
```

Comment:
Under what circumstances is this not redundant with the filtering you do [here](https://github.com/Microsoft/msbuild/pull/1590/files#diff-9d49d0d91290d9f1e0761bdfc60d3533R777)?